### ISSUE TITLE
Ability to apply automatic transformations to text if Forvo lookup fails

### ISSUE THREAD

**Issue by ELLIOTTCABLE** [realmayus/anki_forvo_dl#50]
I'm learning Russian, and a lot of my source-material words are accented to show emphasis (which is important to the pronunciation and for learners, but isn't shown in general text for the most part — and which isn't included on Forvo.)

For instance, the word **прекра́сно** (prikrasna) [yields no results](https://forvo.com/search/%D0%BF%D1%80%D0%B5%D0%BA%D1%80%D0%B0%CC%81%D1%81%D0%BD%D0%BE/), because on Forvo, it's included without the accent: прекрасно [yields 7 results](https://forvo.com/search/%D0%BF%D1%80%D0%B5%D0%BA%D1%80%D0%B0%D1%81%D0%BD%D0%BE/ru/).

It'd be wonderful if there were some sort of way to add automated(ish?) transformations for particular languages — in this instance, applying a script like the following before searching Forvo (or, perhaps, *after* searching Forvo, and then searching with the filtered-version as well):

```javascript
str.normalize("NFD").replace(/\p{Diacritic}/gu, "")
```

**Comment by realmayus** [comment]
Thanks for the issue, I understand the problem. I'm not quite sure how to implement the feature so that it's of use to multiple languages. I'm not familiar with that many languages, but I could imagine that every language might have different needs - e.g. only remove certain diacritics but keep others - so I don't know how to provide a universal option for that.
I think I'll add a "remove all diacritics" toggle on a per-deck basis nevertheless, as these other cases seem rather seldom.
I'm also in the process of implementing a feature allowing users to add RegEx replacement filters that get applied before the query is sent to the server - see #49 for that.

Maybe I could add a section for advanced users (also on a per-deck basis) where they can manipulate a string to their liking in python code, which gets interpreted as a function by the add-on? I'm interested in your thoughts on this!

**Comment by realmayus** [comment]
Also regarding your code snippet, is that just pseudocode? I wasn't able to run it, but I found [this](https://stackoverflow.com/questions/517923/what-is-the-best-way-to-remove-accents-normalize-in-a-python-unicode-string) way of removing all diacritics:
```py
import unicodedata
def strip_accents(s):
   return ''.join(c for c in unicodedata.normalize('NFD', s)
                  if unicodedata.category(c) != 'Mn')
```
But it's quite aggressive and would also remove other characters, since it just filters by the character category `Nonspacing_Mark`.

**Comment by ELLIOTTCABLE** [comment]
Yeah, Unicode normalization is not guaranteed to be information-preserving — be very careful with making that a default behaviour. There's more information available about "when that's a good idea" in the various Unicode Annex documentata … but, basically speaking, 'only normalize when searching, not when storing.'

It's definitely ideal to do the *precise* search first, then issue a second, mutated search IFF the precise content had no matches, right? Or perhaps do both, and merge the results into a single view for the user to select from?

(The code-snippet is, in fact, JavaScript, not Python. I don't know Python very well, I spend all day in OCaml and JavaScript, heh. The normalizations (like "NFD") are defined independant of any implementation, though, and should be universal.)

I doubt full, turing-complete manipulation should be necessary. If someone wants that, and knows enough to need it, they can just … fork the addon? :P
